### ISSUE TITLE
git.latest fails on tag in 2015.2.0rc2

### ISSUE THREAD

**Issue by Bjwebb** [saltstack/salt#22514]
Contents of gitbug.sls:
``` sls
https://github.com/Icinga/icingaweb2.git:
  git.latest:
    - rev: v2.0.0-beta3
    - target: /root/icingaweb2/
```

The first run succeeds, but applying the state again leads to an error. In my limited testing, this error occurs only for git tags and not branches or commits.

```
$ salt-call --local state.sls gitbug
[INFO    ] Loading fresh modules for state activity
[INFO    ] Fetching file from saltenv 'base', ** done ** 'gitbug.sls'
[INFO    ] Running state [https://github.com/Icinga/icingaweb2.git] at time 20:08:40.879426
[INFO    ] Executing state git.latest for https://github.com/Icinga/icingaweb2.git
[INFO    ] Executing command "git clone https://github.com/Icinga/icingaweb2.git '/root/icingaweb2/' " in directory '/root'
[INFO    ] Executing command 'git checkout  v2.0.0-beta3 ' in directory '/root/icingaweb2/'
[INFO    ] Executing command 'git rev-parse HEAD' in directory '/root/icingaweb2/'
[INFO    ] Repository https://github.com/Icinga/icingaweb2.git cloned to /root/icingaweb2/
[INFO    ] {'new': 'https://github.com/Icinga/icingaweb2.git', 'revision': 'ec60f15f09e381f4c622088f16918bfb88e17fa0'}
[INFO    ] Completed state [https://github.com/Icinga/icingaweb2.git] at time 20:08:49.320007
local:
----------
          ID: https://github.com/Icinga/icingaweb2.git
    Function: git.latest
      Result: True
     Comment: Repository https://github.com/Icinga/icingaweb2.git cloned to /root/icingaweb2/
     Started: 20:08:40.879426
    Duration: 8440.581 ms
     Changes:   
              ----------
              new:
                  https://github.com/Icinga/icingaweb2.git
              revision:
                  ec60f15f09e381f4c622088f16918bfb88e17fa0

Summary
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
$ salt-call --local state.sls gitbug
[INFO    ] Loading fresh modules for state activity
[INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://gitbug.sls'
[INFO    ] Running state [https://github.com/Icinga/icingaweb2.git] at time 20:08:52.893053
[INFO    ] Executing state git.latest for https://github.com/Icinga/icingaweb2.git
[INFO    ] Executing command 'git rev-parse HEAD' in directory '/root/icingaweb2/'
[INFO    ] Executing command 'git rev-parse --abbrev-ref HEAD' in directory '/root/icingaweb2/'
[ERROR   ] 'NoneType' object has no attribute 'startswith'
[INFO    ] Completed state [https://github.com/Icinga/icingaweb2.git] at time 20:08:52.904348
local:
----------
          ID: https://github.com/Icinga/icingaweb2.git
    Function: git.latest
      Result: False
     Comment: 'NoneType' object has no attribute 'startswith'
     Started: 20:08:52.893053
    Duration: 11.295 ms
     Changes:   

Summary
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
```

I think this issue may be related to https://github.com/saltstack/salt/issues/22056#issuecomment-91066465

**Comment by jfindlay** [comment]
@Bjwebb, thanks for reporting.

**Comment by Bjwebb** [comment]
Yes, this happens consistently. If I remove the repository. I can reproduce the entire process (with the first run succeeding, and the second run failing), and also every run after the second fails in the same way as the second run.

Output of `salt-call -l debug --local state.sls gitbug` https://gist.github.com/Bjwebb/c9676fa89a2e2e668fe4

Also, here's the output of the same command is run the first time (when the repository does not yet exist) https://gist.github.com/Bjwebb/1ca1eb0ab50bd6578487

**Comment by jfindlay** [comment]
@Bjwebb, thanks for the extra info.

**Comment by terminalmage** [comment]
The git.latest state has been rewritten in https://github.com/saltstack/salt/pull/26341 and now works great with tags. Previously it had odd behavior in certain cases, like prepending the remote name when trying to git reset to the tag, but those issues are now cleared up.

The fix will be in version 2015.8.0, which will be released soon.

Marking this issue as closed.
